### PR TITLE
Explosive birth of automatic casts!

### LINKED ISSUE #2995: Untyped number literals
I suggest that number literals like `7` should not be immediately bound to the type `Int32`, but instead they should be implicitly convertible to any `Number`. Similarly, fractional number literals like `3.43` should be implicitly convertible to any `Float`. To be more exact, no "conversion" would ever take place, ideally the literals would be untyped until they're actually used.

In situations where a type is required (e.g. when writing `a = 0`) the literals default to `Int32`/`Float64` like they used to.
Typed literals like `1.2f32` keep working like they used to.

This change is expected to be backwards compatible, just leads to less verbose code and more permissive compilation &mdash; with no downsides (other than compiler complexity), in my opinion.

---

In other words, I'm sick of these errors:

``` crystal
def shoot(x : Float32, y : Float32)
end

shoot(1.23, 4.56)
# no overload matches 'shoot' with types Float64, Float64
```

``` crystal
record Point, x : Float64, y : Float64

Point.new(0, 0)
# no overload matches 'Point.new' with types Int32, Int32
```

``` crystal
record Color, r : UInt8, g : UInt8, b : UInt8, a : UInt8 = 0

Color.new(255, 128, 0)
# no overload matches 'Color.new' with types Int32, Int32, Int32

Color.new(255u8, 128u8, 0u8) 
# still broken, can you see why?
# instance variable '@a' of Color must be UInt8, not Int32
```

This change would make these examples work.

### LINKED ISSUE #6067: Feature Idea: Auto-prefix `class` and `enum` arguments.
I am interested in getting a little more involved with the actual compiler and have an idea which I think would be a big improvement to the use of `enum`s. If its thought to be a good idea I would appreciate some guidance on where it might be best to start.

Currently when one submits an enum as an argument it needs to be something like `Process::Redirect::Close`. It could simply be `Close`. The method is expecting a `Process::Redirect`. This produces long lines of repeated information; `Process.run(...)` for example.

I would like to modify the compiler to assume a prefix on a given `enum` or `class` as being that of the expected receiving kind. 

Thoughts?

### PR DESCRIPTION
**Pull request by asterite**
## What

This PR lets the compiler match number literals against type restrictions that don't exactly match the expected type, but since the literal's value fits the type, it is automatically cast.

Additionally, a similar behaviour is provided to specify enum values by using a symbol.

All of this works as long as the call doesn't become ambiguous. The implementation was a bit tricky, but finally, and amazingly, it worked.

For a quick understanding of the first part (numbers), you can read what's being requested in #2995. With this PR, **all** of those snippets now compile.

But I'll write a few snippets here.

#### Numbers

```crystal
def foo(x : Int8)
end

foo(1)    # OK, same as passing 1_u8, because 1 fits Int8
foo(1234) # Error
```

```crystal
def foo(x : Float64)
end

foo(1) # OK, same as passing 1.0
```
#### Symbols -> Enums

```crystal
enum Color
  Red
  Green
  Blue
end

def paint(color : Color)
end

paint(:red) # OK, same as passing Color::Red
paint(:yellow) # Error
```
### Instance variables

```crystal
class Foo
  @x : Int8

  def initialize
    @x = 0 # OK, because @x's type is specified, and 0 can pass as 0_u8
  end 
end
```

### Class variables

```crystal
class Foo
  @@x : Int8 = 0 # OK
end
```

### Local variables

```crystal
x : Int8
x = 0 # OK
```

### Method default values, with restriction

```crystal
enum Color
  Red
  Green
  Blue
end

def paint(color : Color = :red) # OK
end

paint # OK
```

### Combined

```crystal
record Vector, x : Float64, y : Float64, z : Float64

# Works because record will use those restrictions in initialize
Vector.new(1, 2, 3)
```

#### File

```crystal
file = File.open(__FILE__)
file.seek(-10, :end) # Same as passing IO::Seek::End
```

#### Process

```crystal
Process.run("ls", output: :inherit) # Same as passing Process::Redirect::Inherit
```

#### String

```crystal
# Well, this one doesn't work, it's missing the type restriction in the method, but it's super easy to fix
"Súper".upcase(:ascii) # => "SúPER"
```

### What if it's ambiguous?

```crystal
def foo(x : Int8)
end

def foo(x : Int16)
end

foo(1) # Error: ambiguous call matches both Int8 and Int16
```

### When it doesn't work

```crystal
def foo(x : Int8)
end

foo(rand < 0.5 ? 1 : 2) # Error
```

All of the above works **only** when passed directly to a method, or assigned directly to an instance/class variable. When putting intermediary expressions, or intermediary assignments, it stops working. But I think that's OK. We simply must document these rules. I also believe for example in Go you can pass number literals like that, but of course it stops working when you introduce intermediate expressions too.

## Why

I can easily say this might be one of the most wanted features in Crystal. Having to type those number suffixes for no real reason, when the compiler could figure this out, or having to type long enum values, is super tedious, and kind of goes against Crystal's philosophy.

Converting symbols to enums might be controversial.

The important thing to understand here is that this is all sound and type safe. If I pass a symbol to a method expecting an enum, and the name can be matched (currently `underscore` is being invoked on both ends), then there's simply no confusion: the symbol must mean that enum member. If there's a typo in the symbol you get an error, exactly like what happens with the full enum member. The big advantage is that you get to type less, and also read less redundant information.

With the rule for numbers you can more easily write numeric and scientific code, and also code for games.

With the rule for symbols -> enum, one is actually more encouraged to use enums, because using them in APIs, when having to choose one option between many, one can still pass a symbol (but you'll have to type the full enum value if choosing the value dynamically, but that's as bad as now).

## How

Implementing this was tricky.

First I started with number literals. My first idea was to give them a different type that would be compatible with other number types if the literal's value fits. That worked, but then this happened:

```crystal
def foo(x : Int8)
end

def foo(x : Int16)
end

def foo(x : Int32)
end

foo(1) # Now matches all of them and ends up calling the first one
```

So I thought I could detect the ambiguous case, and that didn't work at all, because `foo(1)` still matches all overloads.

The solution then came: first try to match methods the old, usual way. If that matches, then all is good and should work like before. If no match is found, only then try again with this special types for literals. If just one matches (or, well, multiple overloads can match, but all of them must match against the same integer type... think about union types in other positions and how that can create a multiple dispatch), then we are good to go. If more than one matches (against different integer types), then it's an ambiguous call.

This is not the most efficient way to do it because when the rule applies, we end up doing two method lookups instead of one. But I think the number of places in a program where this will be used will be much less than the total number of calls where this rule won't be used. And, in any case, this just adds a small performance cost but compilation speed isn't important at this point.

After that, adding the symbol -> enum rule was simple.

Then I had to make sure the rule worked in all of the cases I documented above, which was a bit tedious, but it finally worked.

Fixes #2995
Fixes #6067

### CODE REVIEW

**Comment by S-YOU** [comment]
This is great! Now C bindings code more natural to write.

File: spec/compiler/semantic/automatic_cast.cr [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+      )) { int64 }
+  end
+
+  it "casts literal integer (Int64 -> Int32, ok)" do
+    assert_type(%(
+      def foo(x : Int32)
+        x
+      end
+
+      foo(2147483647_i64)
```
Comment (bew) [inline spec/compiler/semantic/automatic_cast.cr]: Why does this work?
I would expect that if the literal is explicitely typed, that type should be honored

Comment (asterite) [reply spec/compiler/semantic/automatic_cast.cr]: Yeah, maybe that shouldn't work, but I don't know, there's no harm in casting the value.

Plus, right now there's no concept of a number that's **explicitly** typed. Or, well, there's no difference between `1` and `1_i32`, both parse to exactly the same thing, and are handled in the same way in the compiler. So maybe we should start making that distinction and only apply the rule for values that don't have a suffix...

It should be easy to implement, I might try that tomorrow. But I don't think this particular thing should block the entire feature from being merged.

Comment (S-YOU) [reply spec/compiler/semantic/automatic_cast.cr]: Seems okay as long as type on a method or C fun is honored.

**Comment by sdogruyol** [comment]
Welcome back @asterite :tada: :man_dancing:

**Comment by RX14** [comment]
I have to admit, I danced around the room a lot when I saw this pr. 


My only concern is syntax: reusing symbol syntax for enum shorthand vs keeping type syntax and just omitting the full path to the enum type. I'm conflicted. I think a newcomer would see symbols being converted to an enum and be confused.

**Comment by bararchy** [comment]
@RX14 I too was weirded out by the symbols, but paint(Red) makes you think about name spacing issue

**Comment by asterite** [comment]
@RX14 Yeah, that's the only concern about symbol -> enum: confusion. Maybe we can have a separate syntax... or we could simply remove symbols from the language and just use them for this.

But given that there's no ambiguity with this syntax, maybe it's fine.

Well, there's the "ambiguity" of having a method receiving both a Symbol and an Enum, but in that case the Symbol overload wins, of course. But I don't think anyone will write such code...

The other proposal about using `Red` instead of `:red` did have issues as `Red` could be resolving to an existing type other than the enum member.

And then, in Ruby it's very common to pass a symbol as an option. `"FOO".downcase(:ascii)` is also a thing in Ruby, and now we can also have that nice functionality and syntax in Crystal too... Except that in Ruby this even works when passing the symbol through a variable. And maybe **that** will confuse users, and I can imagine questions about this popping all the time in the future. But if it's well documented, you point them to the docs, they read it, and done. Plus it simply won't compile when passed through a variable, so there's no harm other than "Why do I get this error now? Oh, I see, got it".

**Comment by RX14** [comment]
My preference would be for `Type` style instead of `:symbol` style if we can sort out the conflict issues. I like how it's a direct reduction of the enum syntax instead of moving to another syntax (i'd rather not overload symbol syntax).

**Comment by asterite** [comment]
Yeah, `Type` is another option, but it can get confusing if `Type` is also declared as a constant somewhere, and it happens to have the value of an enum member.

Plus I wouldn't know how to implement it. To solve a call, all of its arguments must be typed. But if `Type` doesn't exist in the current scope, you'll get an error before even trying to lookup method matches. Then we have to somehow give `Type` some type (but only if it's in a call position, or in an assignment), delay the error, still do the lookup, and if that doesn't match an enum type we must tell the user that `Type` isn't a type. And I'm not even sure that will work (as I said, I have no idea how to implement this).

Another option is to do something similar to what Swift does: prefix it with a dot. So:

```crystal
"foo".upcase(.Ascii)
```

The type of `.Ascii` will be some kind of `EnumMemberType` that would match an enum type that has that member...

But I don't know, I still prefer the symbol syntax.

**Comment by asterite** [comment]
Cool, this PR also makes the following work (maybe still a bit controversial for enums, but type safe 😛):

```crystal
enum Color
  Red
  Green
  Blue
end

a = [] of Float64
a << 1
a << 2
p a # =>  [1.0, 2.0]

b = [] of Color
b << :red
b << :green
b << :blue
p b # => [Red, Green, Blue]

h = {} of Color => Float64
h[:red] = 1
p h # => {Red => 1.0}
```

Hey, it even makes this work:

```crystal
# All of these currently give an error
[1, 2, 3] of Float64
[1, 2, 3] of UInt8
{1 => 2} of Float64
Set(Float64){1, 2, 3}
```

That's because the code is expanded to a series of calls.

And of course...

```crystal
enum Color
  Red
  Green
  Blue
end

[:red, :green, :blue] of Color
```

This feature is definitely more powerful than the snippets I wrote in the beginning of this PR.

**Comment by Sija** [comment]
IMO there's way less ambiguity with `:symbol` notation than `Type` (and looks better :)).

**Comment by paulcsmith** [comment]
I like the symbol syntax with the same casing as the enum. Makes it super clear this is no "ordinary" symbol :) But I could also be ok with ordinary snake_case symbols as well. Either one is a huge improvement IMO.

A few ideas on making it easy for people to figure out what went wrong when using a symbol:

* Show a "did you mean? :red" when you misspell an enum but it is close to matching one. 
* If they pass a var instead of a SymbolLiteral: "You passed #{what_they_passed}, but you must pass a symbol literal for the 'color' argument. Example: #{first_symbolized_enum}"

I think this would help eliminate some of the confusion that will surely come up and will stop people from banging their heads too long on typos

File: src/compiler/crystal/codegen/cast.cr [hunk 2]
```diff
@@ -488,6 +488,39 @@ class Crystal::CodeGenVisitor
...
+    symbol = @symbols_by_index[index].underscore
+
+    to_type.types.each do |name, value|
+      if name.underscore == symbol
+        accept(value.as(Const).value)
+        return @last
+      end
+    end
+
+    raise "Bug: expected to find enum member of #{to_type} matching symbol #{symbol}"
```
Comment (paulcsmith) [inline src/compiler/crystal/codegen/cast.cr]: Some ideas for this error message: https://github.com/crystal-lang/crystal/pull/6074#issuecomment-387424969

Could easily be done in a future PR if you'd prefer to get this in ASAP

**Comment by RX14** [comment]
What happens in the case of
```cr
enum Foo
  Foo
  FOO
end
```

what does `:foo` match?

This, and readability, makes me want to make the symbols case sensitive.

**Comment by S-YOU** [comment]
```
enum Foo
  Foo
  FOO
end
```

I would prefer FOO automatically cast to Foo, and Foo and FOO should raise already declared error.

**Comment by asterite** [comment]
@RX14 Note that we already define query methods for enums, so there's already a conflict in that snippet (`foo?` will probably return true for `FOO` and not for `Foo`). We can probably disallow enum members whose `underscore` representation turns out to match that of previous enum members.

After that, there's no conflict about using the underscore version or the equal fersion. We still have:

```crystal
module Color
  Red
  Green
  Blue
  CoolOne
end

color = Color::Red
color.red? # => true
color.cool_one? # => false
```

So if we have `red?` and `cool_one?` I see no problem in having `:red` and `:cool_one` for symbols. And I find underscore symbols like that easier to read and write.

We could go with `:Red`, and `:CoolOne`, I don't know... I don't find them as cute. Lower case symbols are very common in Ruby and I'm just used to that. We can have almost the same in Crystal too. 

> A few ideas on making it easy for people to figure out what went wrong when using a symbol

This is not trivial to implement. Maybe even harder than doing this whole PR because it involves heuristics. There might be multiple overloads that didn't match, and then we' have to check if any of the restrictions solve to an enum (maybe the restriction also involves other types in a union), and so on. I'd like to leave this for a separate PR.

**Comment by straight-shoota** [comment]
> Lower case symbols are very common in Ruby and I'm just used to that. We can have almost the same in Crystal too.

That's why I'd like to have some slightly different visuals, to express that the argument is not just a regular symbol but in fact the short notation of an enum member.

**Comment by RX14** [comment]
OK i'm convinced about the problem already existing for the query methods, so there's no technical reason for `:foo` vs `:Foo`. The question is should we be trying to make them visually distinct from normal symbols or not?

**Comment by asterite** [comment]
The way I see it, symbols have very few use cases, enums are almost always better. And mixins APIs with symbols and enums is not common. In fact, I believe not a single public API method in the standard library uses symbols, at all. So I don't see a conflict of using symbols for this.

Again, we could go as far as removing symbols from the language, though that might not be necessary.

**Comment by paulcsmith** [comment]
I personally think that isn't much more to type and is far less "magical". It's incredibly clear 👍

**Comment by luislavena** [comment]
Hello @asterite, great to see you back and sending this 💥of PR! (emoji explained: explosive!)

> But then, I'd like to get rid of the enum question methods, because they use the "undercase" variant. It would be nice to always have to use the member name.

I'm not sure about that, considering is less typing and sometimes enough clarity around the question methods:

```crystal
today = Time.now
if today.friday?
  puts "TGIF!"
end
```
vs:

```crystal
if today.is?(:Friday)
```

```crystal
today.friday?
today.is?(:Friday)
```

Excluding the `.` (dot), one is shorter (7) vs 12 characters. While `.is?` can convey some explicitness, `.friday?` is clear enough too.

Please note that the comment around how much typing is not just about amount of characters that needs to be added, but also the noise it adds around the expression being evaluated. While the enum member is now following the case as shown in the enum itself, they are not the same, and they are hid by the noise of `.is?(...)` around it.

For the example of `case`, perhaps case equality (`===`) could be explored as an alternative instead?

```crystal
color : Color
color = :Red

case color
when :Red
  puts "Red"
when :Blue
  puts "Blue"
end
```

Thank you in advance for your work exploring this solution to make enums more friendlier. ❤️ ❤️ ❤️ 

Cheers.

**Comment by paulcsmith** [comment]
The reason I think `is?` is much clearer than the `?`, is that you can look it up in the docs much more easily, it is easier to see where it comes from, and it *looks* like how you'd use it in other cases so you may not even need to look it up. For those reasons I would prefer `is?`  even though it is a tad bit longer

I agree that having a shorthand for `case` as @luislavena mentioned would be nice! It could come as a separate PR later maybe

**Comment by asterite** [comment]
@luislavena Hi! Thanks for your comment.

What "bothers" me, from a compiler writer, is that one enum question method is defined for each enum member, always. And that adds up to the memory used by the compiler (even if you never invoke that question method). Alternatively I guess we could create those methods dynamically as an optimization, so it's not that bad.

Case equality (`===`) can't be used because you can do `a === b` between any two types, so a symbol will always match that (and the lookup won't enter the phase of "ok, this failed, let's see if the symbol can match an enum member"). Plus, case equality is the other way around. This:

```crystal
case color
when :Red
  something
end
```

is equivalent to:

```crystal
if :Red === color
  something
end
```

and the symbol -> enum rule doesn't work in a call's receiver place.

**Comment by j8r** [comment]
What about `#match` and its sign `=~` instead of `#is?` ? They aren't implemented (always returns `nil`) for  `Symbol`.

**Comment by straight-shoota** [comment]
If enum member arguments are matched case-sensitive, this doesn't have to mean anything for predicate methods. There is no obvious inconsistency in using `foo(:Foo)` and `foo.foo?`. 
Of course we can think about changing these methods and possible alternatives. 

Let's make automatic casts be case-sensitive and revisit predicate methods in a separate issue.

**Comment by asterite** [comment]
We can actually allow method names to start with an uppercase letter. That is already allowed in lib funs, we could extend it to regular methods. Then the query methods can be called exactly like the enum member. So we'll have `:Red` for passing/assigning, and `.Red?` for querying.

**Comment by straight-shoota** [comment]
I don't see a real reason to align the case of enum member names with method names. They're completely different syntactical features.

`:Foo` is essentially a short cut for `My::Enum::Foo` and it makes much more sense to stick with the original notation.

**Comment by luislavena** [comment]
> I see we're getting off-track already.

We are discussing the details of this change and the ripple impact it might have in other areas of usage, like case statements and the potential to make reference to Enum members uniform across the board.

Apologies for being blunt but, how is that off-tracking?

Perhaps the entire point brought up by Ary around question methods and replacement for `.is?` should be discussed on a separate PR, but wasn't me bringing that up, I wanted to raise certain concerns about that point, which definitely deserves a separate and deeper conversation.

> Query methods are instantiated only once, they won't eat much memory or CPU, it's entirely the wrong thing to optimise (please prove me wrong with data :P)

Ary stated that [here](https://github.com/crystal-lang/crystal/pull/6074#issuecomment-387484808). His concerns are from the compiler point of new, not execution.

Cheers.

**Comment by RX14** [comment]
> Apologies for being blunt but, how is that off-tracking?

I meant only the chances to the autogenerated enum query methods, which are definitely off-topic for this PR.

**Comment by asterite** [comment]
Let's do some voting, to see what we all think about our current options:

- 👍 Use this PR as is (`:red`, `.red?`)
- 👎 Change this PR to exactly match symbol to enum member, leave query methods as they are now (`:Red`, `.red?`)
- 😄 Change this PR to exactly match symbol to enum member, but also change query methods to match enum members (`:Red`, `.Red?`)

The last point requires a language change, but the limitation of methods starting with a lowercase letter is artificial. Well, the only problem is top-level methods conflict with types, and of course you wouldn't be able to call such methods without a receiver. But I don't expect uppercase methods to be used a lot, and using them for enum members and lib funs is probably a good use case. Oh, and it's also a breaking change.

**Comment by paulcsmith** [comment]
Can we add a 🎉 For `:Red` and `is?(:Red)`? If not that's ok. Just seeing if that is still on the table :)

**Comment by asterite** [comment]
@paulcsmith done

**Comment by straight-shoota** [comment]
@asterite s/last point/third point/

**Comment by straight-shoota** [comment]
I don't think the specific implications of all the different proposals regarding enum predicate methods have been explored to an extent any decision could be based upon. So I'd go for all but the first option. because the other all equally uninformed.

I'd rather ask for `:red`, `:Red` or `:red unless predicate querying is changed`. The latter means to dive into evaluating the options for predicate methods before continuing with this PR.

**Comment by asterite** [comment]
One more thing about the third (😄) option. We actually don't need to change the language to allow method **def**initions with the first letter being uppercase. Method **calls** already allow this, to be able to invoke lib funs that are named like that.

So, the compiler could create these query methods for enums, even if a user can't. And this means that a user can't override their meaning either, which in my opinion is good. And this can be used in `case` to check for completeness, and again we can be sure nobody can override those methods.

Basically:

```crystal
enum Color
  Red
  Green
  DarkBlue
end

color : Color
color = :DarkBlue

puts color.DarkBlue? # => true

case color
when .DarkBlue?
end
```

Plus these query methods can't conflict between each other like the `underscore`/`snake` versions do.

Easily greppable and consistent everywhere.

(but it's a small breaking change)

**Comment by asterite** [comment]
Just tried the `.DarkBlue` thing locally and it just looks weird, seeing an uppercase letter after a dot. Then there's the `File::Type` enum that defines a `device?` method which is `character_device? || block_device?` and with that change it will look inconsistent.

So my vote finally goes to the first option (👍).

**Comment by straight-shoota** [comment]
I still don't get the drift for equality with predicate methods. Why is that more important than fully qualified type paths?

**Comment by RX14** [comment]
Hopefully we can start a PR of places where we can refactor to use this new syntax to be merged after the next release.

**Comment by oprypin** [comment]
Hmm, not so fast

> This is not the most efficient way to do it because when the rule applies, we end up doing two method lookups instead of one. But I think the number of places in a program where this will be used will be much less than the total number of calls where this rule won't be used. And, in any case, this just adds a small performance cost but compilation speed isn't important at this point.

**Comment by oprypin** [comment]
I mean it may affect the compilation speed of the compiler. It may not be that big of a deal, but also changing to use the new syntax is not a big deal. Maybe it is worth doing only if the effect of it is taken into account.

**Comment by RX14** [comment]
I continue to believe that the solution is vastly reducing the amount of code we have to type (instantiations), not reduce the constant factor of each instantiation.

I don't think this PR will make a serious difference to compile times. If I'm wrong (someone can make a PR which uses this compiler change throughout the stdlib/compiler and measure it) we can revert it.

**Comment by asterite** [comment]
Yeah, I don't think this will have a real impact, or be the bottleneck. Right now method instantiation and method body typing is what takes most of the time, method lookup should be pretty fast as it's a hash lookup and possibly a few type restriction matches, which should also be fast. And very little memory is allocated there compared to when a method is instantiated and all type bindings are created.

**Comment by Sija** [comment]
Would be sweet if autocasting would also work in cases below:

```cr
require "logger"

def foo?(level : Logger::Severity)
  # ...
end

# works
foo? :info

# doesn't work & shouldn't
foo? :nonexistant

# doesn't work & should
foo? true ? :error : :info
foo? x = :error
```

Instead, it fails with:

```
Error in line 11: no overload matches 'foo?' with type Symbol
Overloads are:
 - foo?(level : Logger::Severity)
```

If not possible, it would be helpful to tweak compile error message at least, to something more informative and less misleading than the current one.

**Comment by ysbaddaden** [comment]
It also doesn't work with kwargs:

```crystal
Time.new(2015, 6, 26, 0, 0, 0, location: Time::Location::UTC) # => OK
Time.new(2015, 6, 26, 0, 0, 0, location: :utc) # => ERR: no overload found
```

**Comment by luislavena** [comment]
@Sija from the OP:

> All of the above works **only** when passed directly to a method, or assigned directly to an instance/class variable. When putting intermediary expressions, or intermediary assignments, it stops working. But I think that's OK.

The limitation was expressed in the OP, there is lot of complexity in trying to achieve that (also mentioned the current complexity of this).

Cheres.

**Comment by RX14** [comment]
The problem with the error messages is that we're using the same syntax for two ideas: symbols and enum values. I'd still like to discuss ways of removing symbols, to make the `:enum_member` syntax unambiguous, which would allow better error messages.

I think really the main blocker for removing symbols is the design of named tuples...

**Comment by asterite** [comment]
Let's remove symbols, and make named tuples work only with strings.

**Comment by vladfaust** [comment]
Hey @asterite, why don't you create a discussion for removing symbols and leave them for Enums only? I'm sure you're able to bullet some strong points.

**Comment by j8r** [comment]
@vladfaust the point of removing the `Symbol` type is... to remove it totally, I don't see why keeping it specially for Enums.

**Comment by Sija** [comment]
@j8r point is to have a shorthand for specifiying enum members without a need to pass whole path (`:baz` instead of `Foo::Bar::Baz`).
